secret-provider/src/main/scala/io/lenses/connect/secrets/providers/VaultSecretProvider.scala

```diff
@@ -55,126 +62,85 @@ class VaultSecretProvider() extends ConfigProvider with VaultHelper {
 
   // lookup secrets at a path
   override def get(path: String): ConfigData = {
-    val (expiry, data) = cache.get(path) match {
-      case Some((expiresAt, data)) =>
-        // we have all the keys and are before the expiry
-        val now = OffsetDateTime.now()
-
-        if (expiresAt.getOrElse(now.plusSeconds(1)).isAfter(now)) {
-          logger.info("Fetching secrets from cache")
-          (expiresAt, data)
-        } else {
-          // missing some or expired so reload
-          getSecretsAndExpiry(getSecrets(path))
-        }
-
-      case None =>
-        getSecretsAndExpiry(getSecrets(path))
-    }
-
-    expiry.foreach(exp => logger.info(s"Min expiry for TTL set to [${exp.toString}]"))
-    cache += (path -> (expiry, data))
-    data
+    logger.info(" -> VaultSecretProvider.get(path: {})", path)
+    val sec = cache
+      .cachingWithTtl(
+        path,
+        fnGetMissingValue = getSecretValue(path),
+      )
+      .fold(
+        ex => throw ex,
+        ConfigDataBuilder(_),
+      )
+    logger.info(" <- VaultSecretProvider.get(path: {}, ttl: {})", path, sec.ttl())
+    sec
   }
 
   // get secret keys at a path
   override def get(path: String, keys: util.Set[String]): ConfigData = {
-
-    val (expiry, data) = cache.get(path) match {
-      case Some((expiresAt, data)) =>
-        // we have all the keys and are before the expiry
-        val now = OffsetDateTime.now()
-
-        if (
-          keys.asScala.subsetOf(data.data().asScala.keySet) && expiresAt
-            .getOrElse(now.plusSeconds(1))
-            .isAfter(now)
-        ) {
-          logger.info("Fetching secrets from cache")
-          (
-            expiresAt,
-            new ConfigData(
-              data
-                .data()
-                .asScala
-                .view
-                .filter {
-                  case (k, _) => keys.contains(k)
-                }
-                .toMap
-                .asJava,
-              data.ttl(),
-            ),
-          )
-        } else {
-          // missing some or expired so reload
-          getSecretsAndExpiry(getSecrets(path).view.filter {
-            case (k, _) => keys.contains(k)
-          }.toMap)
-        }
-
-      case None =>
-        getSecretsAndExpiry(getSecrets(path).view.filter {
-          case (k, _) => keys.contains(k)
-        }.toMap)
-    }
-
-    expiry.foreach(exp => logger.info(s"Min expiry for TTL set to [${exp.toString}]"))
-    cache += (path -> (expiry, data))
-    data
+    logger.info(" -> VaultSecretProvider.get(path: {}, keys: {})", path, keys.asScala)
+    val sec = cache.cachingWithTtl(
```

Comment:
I might need help here. If the cache does not contain the value, it should call getSecretValue, and if that fails we want an exception no?

The comment I made earlier is to allow the cache to return Option and then do something like:
```
cache.cachingWithTttl().orElse{
  Option(getSecretValue).map{ v=>
      cache.put
      v
  }.getOrElse(throw)
}
```